src/engineio/asyncio_socket.py

```diff
@@ -66,12 +66,32 @@ async def check_ping_timeout(self):
             return False
         return True
 
-    async def send(self, pkt):
-        """Send a packet to the client."""
+    async def send(self, pkt, apply_bp):
+        """
+        Send a packet to the client.
+
+        params:
+            - apply_bp - applies backpressure handling by limiting
+                         the amount of messages that can go into the
+                         queue. This should not be applied to control
+                         messages.
+        """
         if not await self.check_ping_timeout():
             return
-        else:
-            await self.queue.put(pkt)
+        
+        # back pressure enabled. Check if need to limit
+        if apply_bp and self.queue.qsize() >= self.server.back_pressure_size:
+            self.server.logger.warning(
+                    '%s: back pressure applied. Not Sending packet %s data %s',
+                    self.sid, packet.packet_names[packet.packet_type],
+                    packet.data if not isinstance(packet.data, bytes)
+                    else '<binary>')
+            return
```

Comment:
This is just dropping packets, without giving the application any way to know. If the app is expecting a callback, that'll never arrive. I think this makes it impossible to write a robust application.

A more reasonable approach is what I proposed in https://github.com/miguelgrinberg/python-engineio/issues/133#, which is to set a fixed queue size and let the queue block the app if it reaches the maximum size. With this, no packets are lost, and the service is effectively able to tell the app to slow down by blocking it.

Reply:
I tried to design in such a way that only packets that the server is emitting to the client can be dropped. All others, pings and what not, are fine and are pushed into the queue (at least this is the intention).

There are definitely use cases where you would want to drop the packets. In my use case, I'm sending video to a user. I don't care about a callback as the overhead is too much for me and I don't want the server to buffer all these video packets if the client is unable to keep up with the server. This will cause delayed playback. I think it would be better to only drop packets that are the data itself (not control packets as in heartbeats). That's why I defined a methodology checking the qsize() but not actually using a blocking method to deadlock the queue when I call emit(). deadlocking that queue would be dangerous indeed and I'd love to see a way to actually handle that.

Perhaps the terminology to use here is: "drop_packet_limit". But, not disagreeing that a fixed queue size implementation to prevent dropping off packets would be nice.

Reply:
I get that this works for your case, but it is an extremely limited solution. I still don't understand how control messages are let through. I see that Engine.IO control message are allowed, but what happens when the Socket.IO layer calls the `send()` method in Engine.IO? Socket.IO also has control messages such as callbacks, errors, open/close messages. All those are at risk of being dropped as well and that breaks the protocol.

I suggest you look at how the JS Socket.IO implements "volatile"` messages. I would be okay with an implementation that matches of volatile that matches or is close to the reference JS project.

Reply:
Sure I'll have to look into exactly how they define it. If that's what you'd like to see in the project.

I see this more as an advanced option that can only be toggled on in a senario like I'm discussing. Basically, since TCP is the way that it is, I want it to work more closely to a UDP pipe in which it's fine if a packet or two doesn't show up if the client doesn't have the bandwidth. Video data doesn't care bout that.

But cool sounds too complicated to be accomplished in a few evenings after work with this project.